pkg/v1/cli/catalog/catalog.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package catalog
+
+import (
+	"bytes"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	apimachineryjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+
+	cliv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cli/v1alpha1"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/cli/common"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/utils"
+)
+
+const (
+	// catalogCacheFileName is the name of the file which holds Catalog cache
+	// TODO: Use the original catalog file instead of using v2 once the feature is enabled by default
+	catalogCacheFileName = "catalog_v2.yaml"
+)
+
+var (
+	// PluginRoot is the plugin root where plugins are installed
+	pluginRoot = common.DefaultPluginRoot
+)
+
+// ContextCatalog denotes a local plugin catalog for a given context or
+// stand-alone.
+type ContextCatalog struct {
+	sharedCatalog *cliv1alpha1.Catalog
+	plugins       cliv1alpha1.PluginAssociation
+}
+
+// NewContextCatalog creates context-aware catalog
+func NewContextCatalog(context string) (*ContextCatalog, error) {
+	sc, err := getCatalogCache()
+	if err != nil {
+		return nil, err
+	}
+
+	var plugins cliv1alpha1.PluginAssociation
+	if context == "" {
+		plugins = sc.StandAlonePlugins
+	} else {
+		var ok bool
+		plugins, ok = sc.ServerPlugins[context]
+		if !ok {
+			plugins = make(cliv1alpha1.PluginAssociation)
+			sc.ServerPlugins[context] = plugins
+		}
+	}
+
+	return &ContextCatalog{
+		sharedCatalog: sc,
+		plugins:       plugins,
+	}, nil
+}
+
+// Upsert inserts/updates the given plugin.
+func (c *ContextCatalog) Upsert(plugin cliv1alpha1.PluginDescriptor) error { //nolint:gocritic
+	c.plugins[plugin.Name] = plugin.InstallationPath
+	c.sharedCatalog.IndexByPath[plugin.InstallationPath] = plugin
+
+	if !utils.ContainsString(c.sharedCatalog.IndexByName[plugin.Name], plugin.InstallationPath) {
+		c.sharedCatalog.IndexByName[plugin.Name] = append(c.sharedCatalog.IndexByName[plugin.Name], plugin.InstallationPath)
+	}
+
+	return saveCatalogCache(c.sharedCatalog)
+}
+
+// Get looks up the given plugin.
```

Comment:
**nitpick**: The comment is accurate, but I don't find it to be super helpful.

Maybe something like:
```suggestion
// Get looks up the descriptor of a plugin given its name.
```

Maybe the interface's godoc could be updated with this info as well.